### ISSUE TITLE
ST archiver is very difficult to understand

### ISSUE THREAD

**Issue by jgfouca** [E3SM-Project/E3SM#5419]
I recently tried to enhance MPAS' testing so that MPAS components generated history files and discovered how difficult it is to get the st_archiver to handle a variety of use cases and it seems to especially struggle with MPAS' file naming schemes.

The particular use case is that MPAS wants to archive files like hist.am.highFrequencyOutput but we don't want tests like ERS doing internal comparisons with these files because it won't ever pass because these files were not intended to be used that way. However, the st_archiver won't archive something that it can't identify as a history file or restart file so we are compelled to have a hist_file_extension that matches these files. The only reason any MPAS tests are currently passing is because the CIME testing internals aren't finding the suffixed copies of files like hist.am.highFrequencyOutput because they don't match the expect suffixed-file naming convention, so the internal checks (like ERS's base vs. rest) just get silently skipped (which is another problem).

To sum up, my experience in dealing with MPAS/st_archiver interactions left me feeling that the current st_archiver is complex and brittle; in other words, not very healthy or sustainable for the project. My question is whether folks think it might be worth taking a deep dive on this topic to see if we can improve things? On the surface, this seems like a pretty straightforward problem: we want to look at files produced by a run and move files that match a pattern to a directory and renamed. I'm wondering if config_archive.xml is not expressive enough and therefore too much magic needs to happen in CIME/case/case_st_archive.py (1300 lines long!)? Maybe complexity built-up over the years in CIME and its predecessors which we don't really need anymore?

**Comment by rljacob** [comment]
For discussion, can you paste a list of typical MPAS output files?

**Comment by jonbob** [comment]
@rljacob -- I think the biggest challenge is that both the st_archiver and tests like ERS use the same list of files. The MPAS components create different output files based on both namelist and streams settings, and we'd like them all to be archived. However, many of them will fail cprnc comparisons because they contain either multiple time slices (daily output, for example) or global sums.

**Comment by jgfouca** [comment]
@rljacob:

To be clear, this is not really an MPAS issue. They should be free to use any reasonable naming convention and be able to archive any files they want archived and CIME should be able to support it through config_archive.xml.

To answer your question, here's a set of hist/restart files from `ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods`
```
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.cpl.hi.0001-01-12-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.cpl.hi.0001-01-12-00000.nc.base
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.cpl.hi.0001-01-12-00000.nc.base.cprnc.out
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.cpl.hi.0001-01-12-00000.nc.rest
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.cpl.r.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.eam.r.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.eam.rh0.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.eam.rs.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.elm.r.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.elm.rh0.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mosart.r.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mosart.rh0.0001-01-07-00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpaso.hist.am.globalStats.0001-01-01.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpaso.hist.am.highFrequencyOutput.0001-01-01_00.00.00.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpaso.hist.am.highFrequencyOutput.0001-01-01_00.00.00.nc.base
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpaso.rst.0001-01-07_00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpaso.rst.am.timeSeriesStatsMonthly.0001-01-01_00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpaso.rst.am.timeSeriesStatsMonthlyMax.0001-01-01_00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpaso.rst.am.timeSeriesStatsMonthlyMin.0001-01-01_00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpassi.hist.am.regionalStatistics.0001.01.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpassi.hist.am.timeSeriesStatsDaily.0001-01-01.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpassi.hist.am.timeSeriesStatsDaily.0001-01-01.nc.base
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpassi.rst.0001-01-07_00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpassi.rst.am.timeSeriesStatsDaily.0001-01-01_00000.nc
ERS_P64x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods.20230126_134018_gf8sv0.mpassi.rst.am.timeSeriesStatsMonthly.0001-01-01_00000.nc
```

The micro issue is that they want the `*hist.am*` files to archived like history files but not processed by internal test compares like what happen in ERS. The bigger issue IMO is shortcomings in the case_st_archive system.

**Comment by jonbob** [comment]
I agree with @jgfouca. It would make more sense if case_st_archive could be modified so that the files that get tested can be separate from the files that get archived

**Comment by rljacob** [comment]
I agree that sounds like the best approach.  I think the archive system is where the logic to describe files lives.  The testing system then just uses that.

**Comment by jgfouca** [comment]
@rljacob , the case_st_archive.py code is complex. I was thinking it would be a good candidate for a broader refactor.

**Comment by rljacob** [comment]
Should this be a CIME issue?

**Comment by rljacob** [comment]
The compare problem is that some files have different answers on different number of processors?   And so tests that deliberately change the task count will also give false negatives?

**Comment by xylar** [comment]
To add to the discussion, MPAS-Analysis needs the namelists and streams files themselves.  This has caused downstream users a lot of headache because these files are not (cannot be?) archived and sometimes are not saved along with the output.  As a result, MPAS-Analysis simply cannot be used.

More general archiving that can handle these files would really help me (and the users I provide tech support for).  It could be that this isn't a relevant case for this discussion.  If so, I apologize.

**Comment by jgfouca** [comment]
@xylar , it is definitely relevant. CIME is very picky about which files are eligible for archiving and I think a more flexible system would be better.
